Enforce strict memory safety across WebKit

adetaylor opened the pull request:
#### 6b235623ef129c9a5533b1977160cba712d9cf38
<pre>
Enforce strict memory safety across WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=300439">https://bugs.webkit.org/show_bug.cgi?id=300439</a>
<a href="https://rdar.apple.com/162275605">rdar://162275605</a>

Reviewed by Elliott Williams.

Other WebKit configuration files promote some (few) classes of Swift warnings
to errors. These errors were being suppressed in some cases; this removes that
suppression.

The only class of warning being promoted to error is the strict memory safety
warnings; these become errors. This PR therefore also fixes the remaining
strict memory safety issues.

This is the third attempt at removing this suppression. The first attempt
caused trouble due to these safety errors which did not show up on the EWS bots
in use at the time. EWS is now building using SDK versions which should
correctly show up any remaining problems. The second attempt failed because
VisionOS has some remaining code that does not comply with strict memory
safety; the bug numbers for fixing that are mentioned in a comment.

It is worth noting that PAL.xcconfig still overrides the
OTHER_SWIFT_FLAGS variable and thus does not get the benefit of thesse safety
errors. That&apos;s intentional, because PAL still has some strict safety problems
to be resolved, tracked in <a href="https://rdar.apple.com/164560176">rdar://164560176</a>.

Canonical link: <a href="https://commits.webkit.org/302983@main">https://commits.webkit.org/302983@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/96cfcd47cc2d4f132b1e12e869599b83e9650fb3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130817 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3140 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41776 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138243 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e2d6d5e4-5aee-4090-9517-65b349784aad) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3120 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2983 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99666 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d65a8ff3-1979-4def-baaa-b48200399df3) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2240 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117139 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80368 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ad5dbaff-dd9a-4a14-9acd-a10511e265a4) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2163 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81495 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110755 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35771 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140719 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2884 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2601 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108178 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2930 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113471 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108103 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27508 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2206 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31897 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55893 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2952 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66344 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2773 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2973 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2881 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Configurations/CommonBase.xcconfig:
```diff
@@ -109,6 +109,11 @@ WK_LIBCPP_ASSERTIONS_CFLAGS[sdk=watch*10*] = -D_LIBCPP_ENABLE_ASSERTIONS=1;
 WK_SWIFT_MEMORY_SAFETY_FLAGS = $(WK_SWIFT_MEMORY_SAFETY_FLAGS_$(WK_XCODE_BEFORE_17));
-WK_SWIFT_MEMORY_SAFETY_FLAGS_ = -Werror StrictMemorySafety -strict-memory-safety -enable-experimental-feature Lifetimes -enable-experimental-feature LifetimeDependence -enable-experimental-feature ImportNonPublicCxxMembers;
+// Temporarily disable strict memory safety on VisionOS due to rdar://164555610 and rdar://164559261
+WK_SWIFT_MEMORY_SAFETY_FLAGS[sdk=xros*] = ;
```
what about visionOS simulator?

adetaylor replied:
Good spot, thanks

emw-apple reviewed (approved):
xcconfig changes make sense to me, and the rest of the patch seems like reasonable incremental adoption. Thanks for all the bugs you've filed!

adetaylor committed:
Enforce strict memory safety across WebKit
https://bugs.webkit.org/show_bug.cgi?id=300439
rdar://162275605

Reviewed by Elliott Williams.

Other WebKit configuration files promote some (few) classes of Swift warnings
to errors. These errors were being suppressed in some cases; this removes that
suppression.

The only class of warning being promoted to error is the strict memory safety
warnings; these become errors. This PR therefore also fixes the remaining
strict memory safety issues.

This is the third attempt at removing this suppression. The first attempt
caused trouble due to these safety errors which did not show up on the EWS bots
in use at the time. EWS is now building using SDK versions which should
correctly show up any remaining problems. The second attempt failed because
VisionOS has some remaining code that does not comply with strict memory
safety; the bug numbers for fixing that are mentioned in a comment.

It is worth noting that PAL.xcconfig still overrides the
OTHER_SWIFT_FLAGS variable and thus does not get the benefit of thesse safety
errors. That's intentional, because PAL still has some strict safety problems
to be resolved, tracked in rdar://164560176.

Canonical link: https://commits.webkit.org/302983@main
